packages/styletron-engine-snapshot/src/create-render.js

```diff
@@ -0,0 +1,41 @@
+// @flow
+
+import stringify from "json-stable-stringify";
+
+import {
+  type StyleObject,
+  type FontFaceObject,
+  type KeyframesObject,
+} from "styletron-standard";
+
+/*
+Problem, default atomic `Styletron` engine is really smart and meant for produciton.
+
+It creates atomic classes and only creates `classNames` as necessary.
+
+Unfortunately, for snapshot tests you end up with 2 issues:
+
+1) non-debuggable classNames in snapshots ("ae af ")
+2) className generation is non-deterministic. If you create a new component,
+it might break an unrelated snapshot since that component rendering shifts the
+generation for the old snapshot.
+
+This adds a "deterministic" engine that simply returns a `JSON.stringify`-ed
+version of the style object (with aplha-sorted keys). This makes it so that a
+snapshot should not fail **UNLESS** a style applied to the component changes (which you want).
+It also makes it easier to detect **what** exactly changed in the css in the snapshot since we
+actually now render all of the css values (we are essentially almost rendering the styles inline).
+*/
+export default function createRender(kind: string) {
+  return function render(
+    obj: ?(StyleObject | FontFaceObject | KeyframesObject),
+  ): string {
+    if (!obj) {
+      return "";
+    }
+
+    return `${kind}=${stringify(obj, {space: 2})
+      .replace(/["]/g, "")
```

Comment:
Should this escape quotes instead?

I'd imagine a change like this should yield a snapshot change:
```diff
{
  ":after": {
-   content: "Hello world" // Broken
+   content: `"Hello world"`
  }
}
```

Reply:
Since this shouldn't be used in an actual browser (as no styling works), it might be worth considering using `util.inspect`:

```js
const style = {
  zzz: "zzz"
  ":after": {
    content: `"Hello world"`
  },
  aaa: "aaa"
};

util.inspect(style, {depth: Infinity, sorted: true});
// -> "{ ':after': { content: '\"Hello world\"' },\n  aaa: 'aaa',\n  zzz: 'zzz' }"
```

Reply:
In this case I adopted an approach that makes use of [`prettier`](https://prettier.io/), which uses an `AST` to "pretty print" code.

**Why prettier instead of using the [suggested `util.inspect`](https://github.com/styletron/styletron/pull/342#discussion_r335203582)?**
When exploring using `util.inspect` I ran into a number of issues:

1. Using pure `util.inspect` would require adding a strict node `engine` requirement to `>10` on the package in order to support [`sort` option](https://nodejs.org/docs/latest-v10.x/api/util.html#util_util_inspect_object_options)

2. For backwards compat (to Node v8), one option would be to pre-sort the object before passing it to `util.inspect`.
Packages accomplish this usually by recursively recreating the object in alphabetical order, therefore making sure that the insertion order is
alphabetical.

ie
```js
// illustrative purposes only,
// actual implementation would need to recursively sort nested objects
// + any objects inside of arrays, as well as do some sanitization logic
// before calling `Object.keys`
function sortObj(obj) {
  var sortedObj = {};
  Object.keys(obj).sort().forEach(key => {
    sortedObj[key] = obj[key];
  });
  return sortedObj;
}
```
However, this is [not guaranteed by spec](https://stackoverflow.com/a/30919039), though [in practice this shouldn't be an issue in v8](https://github.com/nodejs/node/blob/master/lib/internal/util/inspect.js#L413)

Was originally going to continue with this approach until realizing that pre `v10`, `util.inspect` does not support the `compact` option, which is needed for `util.inspect` to return a properly indented + newlined string.

**Prettier approach**
The prettier approach instead [works in the browser + node](https://prettier.io/docs/en/browser.html).

The idea is to generate a string that is setting a variable equal to the result of sorted `JSON.stringify`,
and then running that result through `prettier` with a number of options to pretty-print the declaration and strip unnecessary quotes.

example:

In
```js
style = {":after": {"content": "\"Hello World\""}, "color": "red", "zIndex": "-1000"}
```

Out
```js
style = {
  ':after': {
    content: '"Hello World"'
  },
  color: 'red',
  zIndex: '-1000'
}
```